langchain_tools_demo/orchestrator/vertexai_function_calling/function_calling_orchestrator.py

```diff
@@ -0,0 +1,158 @@
+from vertexai.preview.generative_models import (  # type: ignore
+    ChatSession,
+    GenerationResponse,
+    GenerativeModel,
+    Part,
+)
+
+from ..orchestrator import BaseOrchestrator, classproperty
+from .functions import assistant_tool, function_request
+
+MODEL = "gemini-pro"
+BASE_HISTORY = {
+    "type": "ai",
+    "data": {"content": "I am an SFO Airport Assistant, ready to assist you."},
+}
+
+
+class UserChatModel:
+    client: ClientSession
+    chat: ChatSession
+
+    def __init__(self, client: ClientSession, chat: ChatSession):
+        self.client = client
+        self.chat = chat
+
+    @classmethod
+    def initialize_chat_model(cls, client: ClientSession) -> "UserChatModel":
+        model = GenerativeModel(MODEL, tools=[assistant_tool()])
+        function_calling_session = model.start_chat()
+        return UserChatModel(client, function_calling_session)
+
+    async def close(self):
+        await self.client.close()
+
+    async def invoke(self, prompt: str) -> Dict[str, Any]:
+        model_response = self.request_chat_model(prompt)
+        print(f"function call response:\n{model_response}")
+        part_response = model_response.candidates[0].content.parts[0]
+        while "function_call" in part_response._raw_part:
+            function_call = MessageToDict(part_response.function_call._pb)
+            function_response = await self.request_function(function_call)
+            print(f"function response:\n{function_response}")
+            part = Part.from_function_response(
+                name=function_call["name"],
+                response={
+                    "content": function_response,
+                },
+            )
+            model_response = self.request_chat_model(part)
+            part_response = model_response.candidates[0].content.parts[0]
+        if "text" in part_response._raw_part:
+            content = part_response.text
+            print(f"output content: {content}")
+            return {"output": content}
+        else:
+            raise HTTPException(
+                status_code=500, detail="Error: Chat model response unknown"
+            )
+
+    def request_chat_model(self, prompt: str):
+        try:
+            model_response = self.chat.send_message(prompt)
+        except Exception as err:
+            raise HTTPException(status_code=500, detail=f"Error invoking agent: {err}")
+        return model_response
+
+    async def request_function(self, function_call):
+        url = function_request(function_call["name"])
+        params = function_call["args"]
+        print(f"function url is {url}")
+        print(f"params is {params}")
+        response = await self.client.get(
+            url=f"{BASE_URL}/{url}",
+            params=params,
+            headers=get_headers(self.client),
+        )
+        response = await response.json()
+        return response
+
+
+class FunctionCallingOrchestrator(BaseOrchestrator):
+    _user_sessions: Dict[str, UserChatModel] = {}
```

Comment:
should this be a class field? Or an object field?

Reply:
hmm, for the current design seems like we are only generating a single Orchestrator class per app initiation. If we will allow switch between multiple Orchestrator per app in the future then I can change it back to class field. I'll move this definition to the init function! Thankyou!